File: iosApp/iosApp/Modules/CurrencyReceive/CurrencyReceivePresenter.swift
// Created by web3d4v on 13/06/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation

enum CurrencyReceivePresenterEvent {
    case dismiss
}

protocol CurrencyReceivePresenter {
    func present()
    func handle(_ event: CurrencyReceivePresenterEvent)
}

final class DefaultCurrencyReceivePresenter {
    private weak var view: CurrencyReceiveView?
    private let interactor: CurrencyReceiveInteractor
    private let wireframe: CurrencyReceiveWireframe
    private let context: CurrencyReceiveWireframeContext
    
    private var items = [CurrencyReceiveViewModel.Item]()

    init(
        view: CurrencyReceiveView,
        interactor: CurrencyReceiveInteractor,
        wireframe: CurrencyReceiveWireframe,
        context: CurrencyReceiveWireframeContext
    ) {
        self.view = view
        self.interactor = interactor
        self.wireframe = wireframe
        self.context = context
    }
}

extension DefaultCurrencyReceivePresenter: CurrencyReceivePresenter {

    func present() {
        view?.update(
            with: .init(
                title: Localized("currencyReceive.title.receive", context.currency.symbol.uppercased()),
                content: .loaded(
                    .init(
                        name: Localized("currencyReceive.qrcode.name"),
                        symbol: context.currency.symbol.uppercased(),
                        address: interactor.receivingAddress(
                            network: context.network,
                            currency: context.currency
                        ),
                        disclaimer: disclaimer
                    )
                )
            )
        )
    }

    func handle(_ event: CurrencyReceivePresenterEvent) {
        switch event {
        case .dismiss:
            wireframe.dismiss()
        }
    }
}

private extension DefaultCurrencyReceivePresenter {
    var disclaimer: String {
        let arg = "\(context.network.name.capitalized) (\(context.currency.symbol.uppercased()))"
        return Localized("currencyReceive.disclaimer", arg)
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/currencyReceive/CurrencyReceivePresenter.kt
package com.sonsofcrypto.web3walletcore.modules.currencyReceive

import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3walletcore.extensions.Localized
import com.sonsofcrypto.web3walletcore.modules.currencyReceive.CurrencyReceiveWireframeDestination.Back
import com.sonsofcrypto.web3walletcore.modules.currencyReceive.CurrencyReceiveWireframeDestination.Dismiss

sealed class CurrencyReceivePresenterEvent {
    object Back: CurrencyReceivePresenterEvent()
    object Dismiss: CurrencyReceivePresenterEvent()
}

interface CurrencyReceivePresenter {
    fun present()
    fun handle(event: CurrencyReceivePresenterEvent)
}

class DefaultCurrencyReceivePresenter(
    private val view: WeakRef<CurrencyReceiveView>,
    private val wireframe: CurrencyReceiveWireframe,
    private val interactor: CurrencyReceiveInteractor,
    private val context: CurrencyReceiveWireframeContext,
): CurrencyReceivePresenter {

    override fun present() {
        updateView()
    }

    override fun handle(event: CurrencyReceivePresenterEvent) {
        when (event) {
            is CurrencyReceivePresenterEvent.Back -> wireframe.navigate(Back)
            is CurrencyReceivePresenterEvent.Dismiss -> wireframe.navigate(Dismiss)
        }
    }

    private fun updateView() { view.get()?.update(viewModel()) }

    private fun viewModel(): CurrencyReceiveViewModel =
        CurrencyReceiveViewModel(
            Localized("currencyReceive.title.receive", context.currency.symbol.uppercase()),
            Localized("currencyReceive.qrcode.name"),
            context.currency.symbol.uppercase(),
            interactor.receivingAddress(context.network, context.currency) ?: "❌",
            disclaimer
        )

    private val disclaimer: String get() {
        val name = context.network.name.replaceFirstChar {
            if (it.isLowerCase()) it.titlecase() else it.toString()
        }
        val symbol = context.currency.symbol.uppercase()
        return Localized("currencyReceive.disclaimer", "$name ($symbol)")
    }
}